### PR TITLE
API: Deprecate renamae_axis and reindex_axis

### PR DESCRIPTION
**Pull request by TomAugspurger**
Closes https://github.com/pandas-dev/pandas/issues/17833

Some notes:

- the positional argument for `rename_axis` is `mapper`, which isn't really descriptive of what it does now. Oh well
- Added a TODO in https://github.com/pandas-dev/pandas/compare/master...TomAugspurger:depr-rename_axis?expand=1#diff-492dc2a862db88ec1f1d80589ce8a7d6R92. That will (presumably) break in the future when we remove reindex_axis. I haven't attempted to understand what's going on there yet. Maybe it'll be easy

### CODE REVIEW

**Comment by TomAugspurger** [comment]
Hmm, looks like I'll have to give `Panel.reindex` a similar treatment as DataFrame.reindex in #17800

**Comment by jorisvandenbossche** [comment]
> Hmm, looks like I'll have to give Panel.reindex a similar treatment as DataFrame.reindex in #17800

I really wouldn't bother with that. Panel is deprecated, and IMO there is no need in making Panel's functions consistent.

File: doc/source/whatsnew/v0.21.0.txt [hunk 1]
```diff
@@ -810,6 +810,8 @@ Deprecations
 - ``.get_value`` and ``.set_value`` on ``Series``, ``DataFrame``, ``Panel``, ``SparseSeries``, and ``SparseDataFrame`` are deprecated in favor of using ``.iat[]`` or ``.at[]`` accessors (:issue:`15269`)
 - Passing a non-existent column in ``.to_excel(..., columns=)`` is deprecated and will raise a ``KeyError`` in the future (:issue:`17295`)
 - ``raise_on_error`` parameter to :func:`Series.where`, :func:`Series.mask`, :func:`DataFrame.where`, :func:`DataFrame.mask` is deprecated, in favor of ``errors=`` (:issue:`14968`)
+- Using :meth:`DataFrame.rename_axis` and :meth:`Series.rename_axis` to alter index or column *labels* is now deprecated in favor of using ``.rename``. ``rename_axis`` may still be used to alter the name of the index or columns (:issue:`17833`).
+- :meth:`~NDFrame.reindex_axis` has been deprecated in favor of :meth:`~NDFrame.reindex`. See :ref`here` <whatsnew_0210.enhancements.rename_reindex_axis> for more (:issue:`17833`).
```
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v0.21.0.txt]: 'NDFrame' is not something that exists in the documentation. You can just use `DataFrame` (it will not be shown because of the `~`, and the docstrings are the same for series/dataframe)

File: pandas/core/computation/align.py [hunk 2]
```diff
@@ -89,6 +89,7 @@ def _align_core(terms):
         for axis, items in zip(range(ndim), axes):
             ti = terms[i].value
 
+            # TODO: handle this for when reindex_axis is removed...
             if hasattr(ti, 'reindex_axis'):
```
Comment (jorisvandenbossche) [inline pandas/core/computation/align.py]: and using 'reindex' instead of 'reindex_axis' is not good enough?

**Comment by TomAugspurger** [comment]
> I really wouldn't bother with that. Panel is deprecated, and IMO there is no need in making Panel's functions consistent.

I wish I had that option :) It seems to be unavoidable though since some generic methods used `reindex_axis` on frame / panel. Changing those to `reindex(foo, axis)` caused failures.

I have it working for Panel. Working on Panel4D now, hopefully not much longer.

**Comment by TomAugspurger** [comment]
We're getting the refactoring of `_validate_axis_style_args` you and Jeff asked for though ;)

**Comment by jorisvandenbossche** [comment]
Another option is to just make a `_reindex_axis` for now that has the actual implementation and is not deprecated, and use that where `reindex_axis` is used now internally, and we can get rid of it once Panel/Panel4D is gone. 
Again, personally I wouldn't loose more time on it than necessary :-)

File: pandas/core/generic.py [hunk 3]
```diff
@@ -913,6 +908,13 @@ def rename_axis(self, mapper, axis=0, copy=True, inplace=False):
...
+        Notes
+        -----
+        Prior to version 0.21.0, ``rename_axis`` could also be used to change
+        the axis *labels* by passing a mapping or scalar. This behavior is
+        deprecated and will be removed in a future version. Use ``rename``
+        instead.
+
         See Also
         --------
         pandas.NDFrame.rename
```
Comment (jorisvandenbossche) [inline pandas/core/generic.py]: sidenote: this is useless (does not make a link). Unless you want to start with shared docstring/substitution, I would just explicitly add both for series/frame: `pandas.Series.rename, pandas.DataFrame.rename`

**Review by jorisvandenbossche — COMMENTED** [review]
some doc comments

File: pandas/core/panel.py [hunk 4]
```diff
@@ -1197,13 +1197,20 @@ def _wrap_result(self, result, axis):
...
         major_axis = (major_axis if major_axis is not None else
                       kwargs.pop('major', None))
         minor_axis = (minor_axis if minor_axis is not None else
                       kwargs.pop('minor', None))
-        return super(Panel, self).reindex(items=items, major_axis=major_axis,
-                                          minor_axis=minor_axis, **kwargs)
+        axes = self._validate_axis_style_args(
+            labels, 'labels', axes=[items, major_axis, minor_axis],
+            axis=axis, method_name='reindex')
+        if self.ndim >= 4:
```
Comment (TomAugspurger) [inline pandas/core/panel.py]: I got a bit annoyed with PanelND :) I think this is acceptable until we remove it.

File: pandas/tests/reshape/test_pivot.py [hunk 5]
```diff
@@ -116,7 +116,9 @@ def test_pivot_table_dropna_categoricals(self):
 
         result_false = df.pivot_table(index='B', columns='A', values='C',
                                       dropna=False)
-        expected_columns = Series(['a', 'b', 'c', 'd'], name='A')
+        expected_columns = (
```
Comment (TomAugspurger) [inline pandas/tests/reshape/test_pivot.py]: I wanted to flag this change in the tests. AFAICT, this is actually the desired behavior? Looking for issues about this now.

File: pandas/core/generic.py [hunk 6]
```diff
@@ -729,6 +730,51 @@ def swaplevel(self, i=-2, j=-1, axis=0):
...
+                out[self._AXIS_NAMES[i]] = value
+
+        aliases = ', '.join(self._AXIS_NAMES.values())
+        if axis is not None:
+            # Using "axis" style, along with a positional arg
+            # Both index and columns should be None then
+            axis = self._get_axis_name(axis)
+            if any(x is not None for x in axes):
+                msg = (
+                    "Can't specify both 'axis' and {aliases}"
```
Comment (jorisvandenbossche) [inline pandas/core/generic.py]: add ". " at the end of this string? (point + space)

File: pandas/core/groupby.py [hunk 7]
```diff
@@ -4580,7 +4580,7 @@ def _get_sorted_data(self):
 
         # this is sort of wasteful but...
         sorted_axis = data.axes[self.axis].take(self.sort_idx)
-        sorted_data = data.reindex(sorted_axis, axis=self.axis)
+        sorted_data = data.reindex_axis(sorted_axis, axis=self.axis)
```
Comment (jorisvandenbossche) [inline pandas/core/groupby.py]: This won't cause a deprecation warning in our own code?

File: pandas/core/generic.py [hunk 8]
```diff
@@ -729,6 +730,51 @@ def swaplevel(self, i=-2, j=-1, axis=0):
...
+                    "\t.{method_name}({arg_name}, axis=axis), or\n"
+                    "\t.{method_name}(index=index, columns=columns)"  # TODO
+                ).format(arg_name=arg_name, method_name=method_name,
+                         aliases=aliases)
+                raise TypeError(msg)
+            out[axis] = arg
+
+        elif _all_not_none(arg, *axes):
+            msg = (
+                "Cannot specify all of '{arg_name}', {aliases}"
```
Comment (jorisvandenbossche) [inline pandas/core/generic.py]: also missing ". "

File: pandas/core/generic.py [hunk 9]
```diff
@@ -729,6 +730,51 @@ def swaplevel(self, i=-2, j=-1, axis=0):
...
+        if axis is not None:
+            # Using "axis" style, along with a positional arg
+            # Both index and columns should be None then
+            axis = self._get_axis_name(axis)
+            if any(x is not None for x in axes):
+                msg = (
+                    "Can't specify both 'axis' and {aliases}"
+                    "Specify either\n"
+                    "\t.{method_name}({arg_name}, axis=axis), or\n"
+                    "\t.{method_name}(index=index, columns=columns)"  # TODO
```
Comment (jorisvandenbossche) [inline pandas/core/generic.py]: does your 'todo' point to the fact that "index=index, columns=columns" is not generic for all NDFrame types ? (like Panels) If that is the case, I personally wouldn't care.

File: pandas/tests/reshape/test_pivot.py [hunk 10]
```diff
@@ -116,7 +116,9 @@ def test_pivot_table_dropna_categoricals(self):
...
+            Series(['a', 'b', 'c', 'd'], name='A').astype('category')
+        )
```
Comment (jorisvandenbossche) [inline pandas/tests/reshape/test_pivot.py]: how is this change related?

**Review by jorisvandenbossche — COMMENTED** [review]
The support for Panel(nd) makes it much less readable .., but since you already did the hard work, it's ok for me :-) 
Added some more comments.

File: pandas/core/groupby.py (hunk 7 above)
Comment (TomAugspurger) [reply pandas/core/groupby.py]: `data` here is a BlockManager (I incorrectly changed this in an earlier commit)

File: pandas/core/generic.py (hunk 9 above)
Comment (TomAugspurger) [reply pandas/core/generic.py]: Yep, I'll happily ignore :)

File: pandas/tests/reshape/test_pivot.py (hunk 10 above)
Comment (TomAugspurger) [reply pandas/tests/reshape/test_pivot.py]: Not sure :) The fact that it's only not a CategoricalIndex when in the columns and dropna=True seems like a bug on master.

```python
# This is on master
In [7]: df.pivot_table(index="B", columns="A", values="C", dropna=False).columns
Out[7]: Index(['a', 'b', 'c', 'd'], dtype='object', name='A')

In [8]: df.pivot_table(index="A", columns="B", values="C", dropna=False).index
Out[8]: CategoricalIndex(['a', 'b', 'c', 'd'], categories=['a', 'b', 'c', 'd'], ordered=False, name='A', dtype='category')
```

I've added a whatsnew.

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jreback** [comment]
https://travis-ci.org/pandas-dev/pandas/jobs/287943551

@TomAugspurger some deprecation warnings still in test_panel I think. can you fix.
